File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.ts
function search(nums: number[], target: number): number {
    let left = 0 
    let right = nums.length -1 

    while (left < right) {
        let mid = Math.floor((left + right) / 2)
        console.log(left,right,mid)

        if (nums[mid] == target) return mid

        if (nums[left] < nums[mid]) {
            if (target == nums[left]) return left
            if (nums[left] < target && target < nums[mid]) {
                right = mid - 1
            } else {
                left = mid + 1
            }
        } else {
            if (target == nums[right]) return right
            if (nums[mid] < target && target < nums[right]) {
                left = mid + 1
            } else {
                right = mid - 1
            }
        }
        console.log(left,right,mid)
    }
    
    return nums[left] == target ? left : -1
};

File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        left = 0
        right = len(nums) - 1

        while left < right:
            mid = (left + right) // 2

            print(left, right, mid, nums[left:right+1])
            if nums[mid] == target:
                return mid
            elif nums[mid] < nums[right]:
                print('# right sorted')               
                if nums[right] == target:
                    return right
                elif nums[mid] < target < nums[right]:
                    print('# between mid and right')  
                    left = mid + 1  
                else:                
                    print('# at left')                   
                    right = mid - 1
            else:                                       
                print('# left sorted')
                if nums[left] == target:
                    return left
                elif nums[left] < target < nums[mid]:  
                    print('# between left and mid')
                    right = mid - 1  
                else:                                  
                    print('# at right')
                    left = mid + 1

        if nums[left] == target:
            return left
        else:
            return -1
            